tests/draft-next/refRemote.json

```diff
@@ -99,7 +99,40 @@
         ]
     },
     {
-        "description": "base URI change - change folder in subschema",
+        "description": "base URI change - change folder in subschema with root ref",
+        "schema": {
+            "$id": "http://localhost:1234/scope_change_defs2.json",
+            "type" : "object",
+            "properties": {
+                "list": {"$ref": "#/definitions/baz/definitions/bar"}
+            },
+            "definitions": {
+                "baz": {
+                    "$id": "baseUriChangeFolderInSubschema/",
+                    "definitions": {
+                        "bar": {
+                            "type": "array",
+                            "items": {"$ref": "folderInteger.json"}
+                        }
+                    }
+                }
+            }
+        },
```

Comment:
Why do I need to pick a new directory name? Tests should be completely independent, so using the same ID between tests shouldn't impact anything.

(`definitions` is a copy pasta mistake.)

Reply:
> How is your implementation finding these, unless it naively looks for all $id everywhere, even inside constructs like enum and examples?

The reference is just a pointer from the root of the schema.  I just follow the path and try to interpret what I find there as a schema. If I can, then I apply it.  (In that way, it also doesn't matter that I'm using `definitions` here, either.)

Reply:
It's come up in the past that we like to have unique identifiers, so the implementation does not have to be torn down and rebuilt for every test. I'm not the only person who has made this request.

`"$id": "baseUriChangeFolderInSubschema/"` won't be recognized as an addressable schema location in drafts 2019-09/2020-12 if it is in `definitions` rather than `$defs`. It needs to be recognized before the `$ref` is evaluated.

Reply:
> so the implementation does not have to be torn down and rebuilt for every test

I'll change this for now, but I think not being able to run two independent validations on different schemas that share an ID reveals a concurrency/caching problem within the implementation.

An implementation should be able to run each validation completely isolated from other validations.  The way I accomplish this is by having multiple schema registries, one global and one per run (local).  The local registry stores any IDs related to the schema being run, while the global one stores meta-schemas.  The user can also manually register schemas in either, and it's clear that schemas stored in the global registry will be available to all validations.

Reply:
Take the following with a grain of salt, as I plan to change the tests to use `$defs` because that what they should be testing.

> `"$id": "baseUriChangeFolderInSubschema/"` won't be recognized as an addressable schema location in drafts 2019-09/2020-12 if it is in `definitions` rather than `$defs`.

Why not?  As I stated, my implementation will find the location and _then_ attempt to interpret as a schema.  Don't we have tests that verify an implementation can `$ref` into random locations?

> It needs to be recognized before the `$ref` is evaluated.

Why?  There's nothing in the spec that says the schemas MUST be identified in a preprocess step.  How would `$ref`ing into random locations work otherwise?

Reply:
(I agree, but @karenetheridge is correct that we've previously accommodated this at their request, and since the spec doesn't mandate what you're saying and it's easy to avoid, that we should just do so. But +1 on it being a good thing for an implementation to not need this.)

Reply:
> How would $refing into random locations work otherwise?

I'm not sure if you mean arbitrary json pointers, or arbitrary URIs. Arbtirary URIs *don't* work unless the implementation is built to know how to access the network or disk (or whatever else the URI scheme uses) in order to load that resource, and this is not required as per the spec. Arbitrary json pointers don't require loading anything - you just start at the fragmentless location and then walk through the document following properties and indexes to reach the location (and, as discussed elsewhere, if this crosses an $id boundary then even thought is not guaranteed to work).

Reply:
> I'll change this for now, but I think not being able to run two independent validations on different schemas that share an ID reveals a concurrency/caching problem within the implementation.

That is an interesting design choice that I'm not sure is supported by the spec, but I'm not interested in debating it here.

Reply:
> I'm not sure if you mean arbitrary json pointers, or arbitrary URIs

The context of this test is pointers.  That's what I'm talking about.

I'm still unclear why walking within a document doesn't need to be supported if it crosses an `$id` boundary.  Nothing is bundled; it's still a single document.  At this point in dereferencing, the document is just JSON.  Why does moving into an object that contains an `$id` property a problem?

Reply:
> I just follow the path and try to interpret what I find there as a schema. If I can, then I apply it. (In that way, it also doesn't matter that I'm using `definitions` here, either.)

This was the intended interpretation. We discussed it for draft-06. You can only use `$ref` to reference a schema and anything that is referenced should be interpreted as a schema. However, this leaves a bit of an edge case with `$id` that I don't think was ever considered.

We also talked about this when `definitions` was renamed to `$defs`. You can name your definitions whatever you want. The only difference is that you don't get meta-schema validation if you use something non-standard. (There's also that `$id` edge case, but I don't think anyone noticed that at that point either.)

Reply:
> this leaves a bit of an edge case with `$id`

But why?  Why does it matter if `$id` is a property of an intermediate object when navigating via a pointer?  No one is explaining this.

Reply:
Is there agreement on what to do here (yet?) with this PR?

Reply:
I requested a few changes:
- fix the directory name so the fully-resolved URI $id is unique;
- rebase away the changes to the format tests.

Reply:
I'm still not clear why the `$id` needs to be different for all of the tests.  They should be run completely independently.  That some implementations can't do that indicates a problem with the implementation IMO.

Reply:
It's not a questions of *can't*, it's a question of efficiency. We do not state that they must be run independently, and it can be a lot faster for an implementation to just load the remotes/ files once rather than for every single test. Others have stated a preference for independent names (to allow for this) as well.